rules.go

```diff
@@ -0,0 +1,218 @@
+		"thirdparty",
+		"webrtc",
+		"websocket",
+		"xmlhttprequest",
+	}
+	// optionsSplitPat    = fmt.Sprintf(",(?=~?(?:%v))", strings.Join(binaryOptions, "|"))
+	optionsSplitPat    = fmt.Sprintf(",(~?(?:%v))", strings.Join(binaryOptions, "|"))
+	optionsSplitRe     = regexp.MustCompile(optionsSplitPat)
+	escapeSpecialRegxp = regexp.MustCompile(`([.$+?{}()\[\]\\])`)
+)
+
+// Structs
+
+type Request struct {
+	// parsed full URL of the request
+	URL *url.URL
+	// a value of Origin header
+	Origin string
+	// a value of Referer header
+	Referer string
+	// Defines is request looks like XHLHttpRequest
+	IsXHR bool
+}
+
+type Rule struct {
+	raw         string
+	ruleText    string
+	rawOptions  []string
+	regexString string
+	options     map[string]bool
+	optionsKeys []string
+	isComment   bool
+	isHTMLRule  bool
+	isException bool
+	document    bool
+	domain      map[string]bool
+}
+
+// Interfaces
+
+type RuleSet interface {
+	Allow(*Request) bool
+}
+
+// Allow methods
+
+func (rule *Rule) Allow(url string) bool {
+	// TODO
+	return false
+}
+
+//
+
+func ParseRule(ruleText string) (*Rule, error) {
+	rule := &Rule{}
+	rule.raw = ruleText
+	ruleText = strings.TrimSpace(ruleText)
+	rule.isComment = strings.Contains(ruleText, "!") || strings.Contains(ruleText, "[Adblock")
+	if rule.isComment {
+		rule.isHTMLRule = false
+		rule.isException = false
+	} else {
+		rule.isHTMLRule = strings.Contains(ruleText, "##") || strings.Contains(ruleText, "#@#")
+		rule.isException = strings.HasPrefix(ruleText, "@@")
+		if rule.isException {
+			ruleText = ruleText[2:]
+		}
+	}
+
+	rule.options = make(map[string]bool)
+	if !rule.isComment && strings.Contains(ruleText, "$") {
+		var option string
+		parts := strings.SplitN(ruleText, "$", 2)
+		length := len(parts)
+		if length > 0 {
+			ruleText = parts[0]
+		}
+		if length > 1 {
+			option = parts[1]
+		}
+
+		rule.rawOptions = strings.Split(option, ",")
+		for _, opt := range rule.rawOptions {
+			if strings.HasPrefix(opt, "domain=") {
+				rule.domain = parseDomainOption(opt)
+			} else {
+				rule.options[strings.TrimPrefix(opt, "~")] = !strings.HasPrefix(opt, "~")
+			}
+		}
+	} else {
+		rule.rawOptions = []string{}
+		rule.domain = make(map[string]bool)
+	}
+
+	rule.optionsKeys = rule.OptionsKeys()
+	rule.ruleText = ruleText
+
+	if rule.isComment || rule.isHTMLRule {
+		rule.regexString = ""
+	} else {
+		var err error
+		rule.regexString, err = ruleToRegexp(ruleText)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return rule, nil
+}
+
+func NewRuleSet(rules []*Rule) (RuleSet, error) {
+	return nil, nil
+}
+
+func (rule *Rule) OptionsKeys() []string {
+	opts := []string{}
+	for opt := range rule.options {
+		opts = append(opts, opt)
+	}
+	if rule.domain != nil && len(rule.domain) >= 0 {
+		opts = append(opts, "domain")
+	}
+	return opts
+}
+
+func parseDomainOption(text string) map[string]bool {
+	domains := text[len("domain="):]
+	parts := strings.Split(strings.Replace(domains, ",", "|", -1), "|")
```

Comment:
I'm not sure if I follow you here. Why do you need to replace all `,` with `|` and split by that symbol? You can split by `,` 🤔

Reply:
Actually it is just split on `|` 🙂